### PR TITLE
Improve Kestrel endpoints

### LINKED ISSUE #21117: Improve Kestrel endpoints page
Giant Kestrel page [is being split up](https://github.com/dotnet/AspNetCore.Docs/pull/21091). After that PR is merged the new standalone endpoints page should be improved.

Currently the most important sections are focused around APIs, rather than what a user wants to achieve.

Possible table of contents:

* Introduction to endpoints (e.g. high-level description of address, protocol, HTTPS, etc)
* Configure endpoints with `URLs`
  * URL prefixes
* Configure endpoints in appsettings.json
  * Configure endpoint defaults
  * HTTPS
* Configure endpoints in code
  * Endpoint defaults
  * HTTPS
  * TCP socket endpoints
  * Unix socket endpoints
  * Dynamic port binding
  * Connection logging
* Configure HTTPS
  * Summary (links to other HTTPS sub-sections)
  * Configure endpoints using SNI
* Endpoint protocols

@halter73 @tratcher

tdykstra edit to add Document details section

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: 312df0ba-e769-c904-40b2-733c291704a0
* Version Independent ID: fd9cc0c2-3ea3-f437-e0ac-8789fb96faf7
* Content: [Configure endpoints for the ASP.NET Core Kestrel web server](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints?view=aspnetcore-7.0)
* Content Source: [aspnetcore/fundamentals/servers/kestrel/endpoints.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/fundamentals/servers/kestrel/endpoints.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-fundamentals**
* GitHub Login: @tdykstra
* Microsoft Alias: **tdykstra**
* 

---
[Associated WorkItem - 82758](https://dev.azure.com/msft-skilling/Content/_workitems/edit/82758)

### PR DESCRIPTION
**Pull request by JamesNK**
Based on https://github.com/dotnet/AspNetCore.Docs/pull/29501
Fixes #21117 

[The issue](https://github.com/dotnet/AspNetCore.Docs/issues/21117) initially proposed grouping topics by ways to configure (e.g. code, appsettings.json). Then under each way to configure, there would be subjects (e.g. HTTPS, HTTP protocol)

That layout didn't work well because subjects couldn't be introduced once (e.g. here are all the HTTP protocols and what is needed to support them) because the instructions to configure the subject was in 2-3 different places. The layout is now by subject, with an introduction to that subject, then ways to configure it as subheadings.

I've liberally added links between different sections, so people can jump directly to the content they want.

**Use this link to preview changes:** https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints?view=aspnetcore-8.0&branch=pr-en-us-29564

For reference, here is the current page: https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints?view=aspnetcore-7.0

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/fundamentals/servers/kestrel/connection-middleware.md](https://github.com/dotnet/AspNetCore.Docs/blob/580e087162762e67791dcd0a7ddaf6353a5a8e50/aspnetcore/fundamentals/servers/kestrel/connection-middleware.md) | [aspnetcore/fundamentals/servers/kestrel/connection-middleware](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/connection-middleware?branch=pr-en-us-29564) |
| [aspnetcore/fundamentals/servers/kestrel/endpoints.md](https://github.com/dotnet/AspNetCore.Docs/blob/580e087162762e67791dcd0a7ddaf6353a5a8e50/aspnetcore/fundamentals/servers/kestrel/endpoints.md) | [Configure endpoints for the ASP.NET Core Kestrel web server](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints?branch=pr-en-us-29564) |

### CODE REVIEW

**Comment by JamesNK** [comment]
@tdykstra I made **_a lot_** of changes.

I refactored your PR to change the layout. I proposed grouping topics by method (i.e. code, appsettings.json) in [the issue](https://github.com/dotnet/AspNetCore.Docs/issues/21117). I found that layout didn't work well because subjects (i.e. HTTPS, HTTP protocol) couldn't be introduced once because their instructions are spread everywhere. The layout is now by subject.

Can I copy these changes into your PR?

**Comment by JamesNK** [comment]
I have no idea where this section should go. Why was it added?

> On Linux, [CipherSuitesPolicy](https://learn.microsoft.com/en-us/dotnet/api/system.net.security.ciphersuitespolicy) can be used to filter TLS handshakes on a per-connection basis:

```csharp
var builder = WebApplication.CreateBuilder(args);

builder.WebHost.ConfigureKestrel((context, serverOptions) =>
{
    serverOptions.ConfigureHttpsDefaults(listenOptions =>
    {
        listenOptions.OnAuthenticate = (context, sslOptions) =>
        {
            sslOptions.CipherSuitesPolicy = new CipherSuitesPolicy(
                new[]
                {
                    TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
                    TlsCipherSuite.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
                    // ...
                });
        };
    });
});
```

**Comment by tdykstra** [comment]
@JamesNK 

> Can I copy these changes into your PR?

Yes

**Comment by tdykstra** [comment]
@JamesNK Or maybe it would be easier to close my original PR, rename this one, and continue reviews here?

**Comment by JamesNK** [comment]
Sure, let’s do that.

**Comment by JamesNK** [comment]
@tdykstra I think this is ready for final review. Could you look at the language?

**Review by tdykstra — APPROVED** [review]
@JamesNK Thank you for all the work you put into making this better!

**Comment by JamesNK** [comment]
Merged. Whew. This is a very hard article to get right. I think it is a big improvement. Thanks @tdykstra.

**Review by Tratcher — COMMENTED** [review]
Sorry for late comments...

File: aspnetcore/fundamentals/servers/kestrel/endpoints.md [hunk 1]
```diff
@@ -5,144 +5,235 @@ description: Learn about configuring endpoints with Kestrel, the cross-platform
...
 
-* No configuration
-* Replace the default certificate from configuration
-* Change the defaults in code
+#### Bind to a Unix socket
 
-### No configuration
+Listen on a Unix socket with <xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.ListenUnixSocket%2A> for improved performance with Nginx, as shown in this example:
+
+:::code language="csharp" source="~/fundamentals/servers/kestrel/samples/6.x/KestrelSample/Snippets/Program.cs" id="snippet_ListenUnixSocket":::
```
Comment (Tratcher) [inline aspnetcore/fundamentals/servers/kestrel/endpoints.md:193]: The example shows http and https using the same sock? How would that work?

File: aspnetcore/fundamentals/servers/kestrel/endpoints.md [hunk 2]
```diff
@@ -68,81 +126,125 @@ By default, Kestrel configuration is loaded from the `Kestrel` section and reloa
...
-* `UseHttps(X509Certificate2 serverCertificate, Action<HttpsConnectionAdapterOptions> configureOptions)`
-* `UseHttps(Action<HttpsConnectionAdapterOptions> configureOptions)`
+* <xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.Listen%2A>
+* <xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.ListenLocalhost%2A>
+* <xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.ListenAnyIP%2A>
+* <xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.ListenUnixSocket%2A>
+* <xref:Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.ListenNamedPipe%2A>
 
-`ListenOptions.UseHttps` parameters:
+When both the `Listen` and [UseUrls](#configure-endpoints-with-urls) APIs are used simultaneously, the `Listen` endpoints override the `UseUrls` endpoints.
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply aspnetcore/fundamentals/servers/kestrel/endpoints.md:174]: I don't know. This was existing content.

File: aspnetcore/fundamentals/servers/kestrel/endpoints.md (hunk 1 above)
Comment (JamesNK) [reply aspnetcore/fundamentals/servers/kestrel/endpoints.md:193]: I believe the second call will error. I'll remove the https line.

That mistake must have been there for years.

**Pull request merged**